### DISCUSSION TITLE
SAC clawback should also emit a burn event

### DISCUSSION THREAD

**Discussion by chowbao** [stellar/stellar-protocol#1589] [Core Advancement Proposals]
# Proposal
Add emitting a `burn` event on SAC clawback

# Changes
```diff
/// Clawback `amount` from `from` account. `amount` is burned in the
/// clawback process.
///
/// # Arguments
///
/// * `from` - The address holding the balance from which the clawback will
///   take tokens.
/// * `amount` - The amount of tokens to be clawed back.
///
/// # Events
///
/// Emits an event with topics `["clawback", admin: Address, from: Address, sep0011_asset: String],
/// data = amount: i128`
+ /// Emits an event with topics `["burn", from: Address, sep0011_asset: String], data = amount:
+ /// i128`
fn clawback(env: Env, from: Address, amount: i128);
```

# Reasoning
[CAP-46-06](https://github.com/chowbao/stellar-protocol/blob/master/core/cap-0046-06.md) defines that a clawback should burn the amount from the clawback but current implementation does not actually emit a `burn` event. It only emits a new `clawback` event (seen in https://github.com/chowbao/stellar-protocol/blob/master/core/cap-0046-06.md#semantics-admin-interface)

**Comment by dmkozh** [comment]
We probably should just clarify the semantics of the `clawback` better and maybe not use the word 'burn' there. `burn` event is there for self-burning the token. `clawback`  is for admin-controlled burns. I'm not sure it's a good idea to duplicate the events now - it will likely cause confusion in the downstream systems and cause them to double-count the burns.

**Reply by chowbao** [reply]
I think that's a good alternative

**Reply by leighmcculloch** [reply]
@chowbao pointed out in [slack](https://stellarfoundation.slack.com/archives/C02B04RMK/p1733848941508479?thread_ts=1733502333.026599&cid=C02B04RMK) that there's only been one clawback in the history of Soroban, and because we're talking about an impact to off-chain systems that are updateable, I don't think we should block fixing this bug on that reasoning. I would go as far to say that this is a bug.

An alternative is to add clawback events to SEP-41, but that'd be going down a path of overloading SEP-41, which doesn't seem appropriate. Custom token contracts may not have a concept of an admin so the idea that burn events must be non-admin burns doesn't translate well to SEP-41 either and SEP-41 doesn't provide that distinction or semantic today.

One other note is that today burn events don't only get emitted from the action of the holder, because the allowance functionality allows anyone with appropriate approval to burn someone else's events, not only admins. So I think the distinction between admin vs non-admin is relatively unimportant, and for any system that still needs to determine that distinction it is relatively easy to marry matching clawback and burn events isolated within the same transaction.

**Reply by leighmcculloch** [reply]
> Custom token contracts may not have a concept of an admin so the idea that burn events must be non-admin burns doesn't translate well to SEP-41 either and SEP-41 doesn't provide that distinction or semantic today.

To add to this comment, this is what SEP-41 says about burn events, which doesn't say admins can't emit them.

https://github.com/stellar/stellar-protocol/blob/5a22e2deaf56d97497ab0827f2613068e5678064/ecosystem/sep-0041.md?plain=1#L200

**Reply by sisuresh** [reply]
So the alternative to emitting `burn` here would be to specify `clawback` in SEP 41, because some custom tokens could implement clawback and emit a non-`burn` event right @chowbao? Is it not acceptable for the SEP to just say any token burn (even `clawback`) should emit a `burn` event and just ignore what the SAC does? Or is the issue that we want to allow SEP to be used to track SAC balances ~~(which won't work due to various reasons)~~.

**Reply by dmkozh** [reply]
> I would go as far to say that this is a bug.

Why do you consider this a bug? The general principle we have been following is 1 token function => 1 event. I don't think we should break that principle. We should just clarify the clawback semantics and try not overloading the term burn. Otherwise, if we follow the logic of 'if a balance has been decreased, then emit burn event', then we'll have to emit a burn event when a transfer happens (because the amount has been burnt from the source account). I think that's unreasonable.

>  So I think the distinction between admin vs non-admin is relatively unimportant, 

Then we should get rid of clawback event completely and use just the burn event. Clawback do seem different to me though, but if no-one cares, we could get rid of them completely.

> for any system that still needs to determine that distinction it is relatively easy to marry matching clawback and burn events isolated within the same transaction.

I've had this point in some neighbor discussion, but I'll repeat it here: let's not introduce any necessity for the event linking in the protocol. It's not straightforward and I'm pretty sure it might cause issues later down the road. Thus far one could simply reproduce the balance for any address by increasing the balance on mints/incoming transfers and decreasing it on burns/clawbacks/outgoing transfers. I don't think we should break this logic.

I would vote for either getting rid of the clawback events completely in favor of burn, or to just clarify the semantics of clawback and burn to make the distinction more obvious.

**Reply by chowbao** [reply]
> some custom tokens could implement clawback and emit a non-burn event right @chowbao?

Yes

> I would vote for either getting rid of the clawback events completely in favor of burn

I would prefer this but that would be unreasonable because SACs currently have existing clawback events right?

**Reply by leighmcculloch** [reply]
> Why do you consider this a bug?

Because SEP-41 says a burn event should be emitted when the token is burned. Clawback burns the token, it no longer lives in any balance, therefore it is burned. If an asset issue planned to undo a clawback, they'd mint, which is the opposite of burning.

https://github.com/stellar/stellar-protocol/blob/5a22e2deaf56d97497ab0827f2613068e5678064/ecosystem/sep-0041.md?plain=1#L200

> 1 token function => 1 event

This won't scale. It's reasonable for a contract to emit additional events such as the clawback event. When folks build custom tokens that emit additional events, they should still emit the SEP-41 events even if there is overlap, otherwise systems that ingest SEP-41 token events will miss balance changes.

**Reply by leighmcculloch** [reply]
> I would vote for either getting rid of the clawback events completely in favor of burn

Iirc the intent with a separate clawback event was because it is such an important event that it was worthy of yelling loudly that it had occurred. (cc @sisuresh if you remember more) It's also easier for applications to display that a clawback has occurred, otherwise applications may incidentally display a burn on its own with zero context, causing confusion since the user had not performed the action themselves.

The issue of linking is going to keep coming up, because contracts will need to innovate outside of the SEP-41 token standard, and applications will want to understand the relationship between events. The protocol doesn't have to solve that problem at the same time though.

**Reply by dmkozh** [reply]
> This won't scale. 

I mean 1 _standard_ event. I think it can scale, given that SEP-41 standard is pretty narrow and concise. It's of course fine to emit non-standard events, and there is no significant risk that they will get confused with the standard events and lead to double-counting (unlike what is suggested here).

> Clawback burns the token, it no longer lives in any balance, therefore it is burned.

I'm still not sure why do we have to define the semantics specifically in terms of burn. If we think distinguishing clawback is important, then why not say that the token may be removed from supply either via burn, or via clawback? There are no burn events at all in ERC-20, so I'm not sure if we there is some strict semantic standard that we have to follow. One can also transfer tokens to a '0-address' and effectively remove them from supply, which would be considered a 'burn' in ERC-20 world, but we won't emit a `burn` event in such case. 

I'm really struggling to see the benefit of making things more complicated here for the sake of following some arbitrary semantics that we've defined; I don't think anything prevents us from updating the semantics to reflect the real state of things (which currently is 'the token can be removed from supply either via `burn`, or via `clawback`, or via `transfer` with locked-down destination'). The proposed update is 'the token can be removed from supply either via `burn`, or via `transfer` with locked-down destination. `clawback` events are purely informational and shouldn't be used for balance tracking, as their amounts are duplicated by the respective `burn` event'. Does that really simplify things? 

> It's also easier for applications to display that a clawback has occurred, otherwise applications may incidentally display a burn on its own with zero context,

FWIW this is likely to happen if we start emitting `burn` alongside with `clawback` if the app is not careful with joining the events.

> The issue of linking is going to keep coming up, because contracts will need to innovate outside of the SEP-41 token standard, and applications will want to understand the relationship between events. 

I don't disagree that the need might arise eventually, but I think that if we can avoid that need, then we should. Especially in something as fundamental as token standard.

**Reply by chowbao** [reply]
>> The issue of linking is going to keep coming up, because contracts will need to innovate outside of the SEP-41 token standard, and applications will want to understand the relationship between events.
>
> I don't disagree that the need might arise eventually, but I think that if we can avoid that need, then we should. Especially in something as fundamental as token standard.

> The protocol doesn't have to solve that problem at the same time though. 

I agree that linking events doesn't have to be solved immediately nor within this discussion/proposal but shouldn't we think about how to link events sooner rather than later? It seems like it is an inevitability but I guess it's a question of whether that is a month from now or 5 years from now. I feel like it'd be sooner (not 1 month but maybe next year) because of the growth of the network and the plans to increase smart contracts (and classic operations/transactions?) per ledger

**Reply by dmkozh** [reply]
> I agree that linking events doesn't have to be solved immediately nor within this discussion/proposal but shouldn't we think about how to link events sooner rather than later?

This is definitely out of scope of this proposal. But for the custom protocols there is really no issue preventing the event linking; events may have arbitrary identifiers that can then be used by the downstream systems for arbitrary linking. If we want some mechanism at the protocol level, that would need a separate CAP with detailed requirements and motivation.

**Reply by chowbao** [reply]
Another alternative (sorry if I missed it if it was already discussed), we could extend the `burn` event so it has a memo identifying it as clawback. This would 
* Adhere to `1 token function => 1 event`
* Get rid of an extra event type (`clawback` event)

**Reply by dmkozh** [reply]
> event so it has a memo identifying it as clawback. 

I like this more than original proposal as well. I'm not sure what exactly should be the format though. In theory we could be as specific as specifying the type of the initiator of the burn (self, address with allowance, token admin), as well as the initiator address (if applicable). IDK if that's useful in practice though.

**Reply by sisuresh** [reply]
Yeah I agree emitting a burn event with clawback info works as well. Changing the format of the current `burn` is more intrusive though because `burn` is already in SEP-41. The `amount` in the `burn` event is not a container unfortunately. Would we put the clawback info into the topic? Or specify a second `burn` event for backwards compatibility?

**Reply by dmkozh** [reply]
Can't we just append a value after `amount`? I think that's acceptable as a protocol change. That also has pretty low disruption potential (someone could be broken, but the fix is very simple).

**Reply by sisuresh** [reply]
Depends on if we're fine with making a breaking change like that. I agree the fix is simple. Existing tokens also wouldn't be able to adapt as well.

**Reply by dmkozh** [reply]
Anything we might do here (besides preserving the status quo and updating the docs) is a breaking change. I think it's acceptable at protocol boundaries, especially from something as underused as clawbacks are thus far.

**Reply by leighmcculloch** [reply]
> If we think distinguishing clawback is important, then why not say that the token may be removed from supply either via burn, or via clawback?

👍🏻 After more thought I'm changing my opinion and think this is reasonable, that SEP-41 can get a `clawback` event, but no function to perform it similar to the mint proposal, and that clawback event should can be described in a way that is optional (without necessarily using that word, since most elements of the standard are ultimately optional). It can say that the event should be interpreted by external systems as a burn with a different cause, and that the event is intended to be useful to tokens that allow for burning by some action that is outside the control of the token holder, where there is value to signal to the community that token value has been reclaimed by the contract or token. Token implementations do not need to use it, but can if it improves the semantics and transparency of the contract's behaviour.

This pattern of continuing to heap on SEP-41 doesn't scale, but the SAC is the default SEP-41 token, so we should avoid significant divergence in their semantics and behaviours. And clawback is a reasonably generic behaviour that custom tokens that don't support it can just never generate an event for.

So I'm in favor of moving ahead with the other proposal instead:
- https://github.com/stellar/stellar-protocol/discussions/1590

**Comment by chowbao** [comment]
Discussion has slowed.

The main conclusion is that we don't want to emit 2 events for a clawback action. But It's unclear what the conclusion for the CAP should be
* Do we want to get rid of `clawback` events?
  * No
* Instead of the `clawback` event do we want to append clawback info to a `burn` event to behave as the "clawback" event?
  * No
* Update CAP-46-06 to say what events actually should be emitted

**Comment by chowbao** [comment]
Closing discussion. Clawback event will not change. Admin will be removed
